Duplicate symbol in `aes-gcm-vaes-avx2.S` on Linux 6.19 with `--enable-linux-builtin`

<!-- Please fill out the following template, which will help other contributors address your issue. -->

<!--
Thank you for reporting an issue.

*IMPORTANT* - Please check our issue tracker before opening a new issue.
Additional valuable information can be found in the OpenZFS documentation
and mailing list archives.

Please fill in as much of the template as possible.
-->

### System information
<!--  add version after "|" character -->
Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| (rolling)
Kernel Version	| 6.19.0
Architecture	| x86_64
OpenZFS Version	| master
<!--
Command to find OpenZFS version:
zfs version

Commands to find kernel version:
uname -r		# Linux
freebsd-version -r	# FreeBSD
-->

### Describe the problem you're observing

Building Linux 6.19 with the latest OpenZFS `master` _built-in_ causes an error because of duplicate symbols.

### Describe how to reproduce the problem

Build latest OpenZFS `master` against Linux 6.19 with `--enable-linux-builtin`.

### Include any warning/errors/backtraces from the system logs


```
# LD      vmlinux.o
  ld.lld -m elf_x86_64 -mllvm -import-instr-limit=5 -z noexecstack -r -o vmlinux.o  -T .tmp_initcalls.lds --whole-archive vmlinux.a --no-whole-archive --start-group  --end-group  ; ./tools/objtool/objtool --hacks=jump_label --hacks=noinstr --hacks=skylake --ibt --mcount --mnop --orc --retpoline --rethunk --sls --static-call --uaccess --prefix=16 --link vmlinux.o
ld.lld: error: duplicate symbol: aes_gcm_enc_update_vaes_avx2
>>> defined at aes-gcm-vaes-avx2.S:1135 (/usr/src/git/linux-stable/arch/x86/crypto/aes-gcm-vaes-avx2.S:1135)
>>>            arch/x86/crypto/aes-gcm-vaes-avx2.o:(aes_gcm_enc_update_vaes_avx2) in archive vmlinux.a
>>> defined at aesni-gcm-avx2-vaes.S:363 (/usr/src/git/linux-stable/fs/zfs/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.S:363)
>>>            fs/zfs/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.o:(.text+0x460) in archive vmlinux.a

ld.lld: error: duplicate symbol: aes_gcm_dec_update_vaes_avx2
>>> defined at aes-gcm-vaes-avx2.S:1138 (/usr/src/git/linux-stable/arch/x86/crypto/aes-gcm-vaes-avx2.S:1138)
>>>            arch/x86/crypto/aes-gcm-vaes-avx2.o:(aes_gcm_dec_update_vaes_avx2) in archive vmlinux.a
>>> defined at aesni-gcm-avx2-vaes.S:900 (/usr/src/git/linux-stable/fs/zfs/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.S:900)
>>>            fs/zfs/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.o:(.text+0xc00) in archive vmlinux.a
make[2]: *** [scripts/Makefile.vmlinux_o:76: vmlinux.o] Error 1
make[1]: *** [/usr/src/git/linux-stable/Makefile:1257: vmlinux_o] Error 2
make: *** [Makefile:248: __sub-make] Error 2
```

--- a/module/icp/algs/modes/gcm.c
+++ b/module/icp/algs/modes/gcm.c
@@ -34,6 +34,7 @@
 #include <modes/gcm_impl.h>
 #ifdef CAN_USE_GCM_ASM
 #include <aes/aes_impl.h>
+#include <modes/gcm_asm_rename_funcs.h>
 #endif
 
 #define	GHASH(c, d, t, o) \

--- a/module/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.S
+++ b/module/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.S
@@ -7,6 +7,7 @@
 
 #define _ASM
 #include <sys/asm_linkage.h>
+#include <modes/gcm_asm_rename_funcs.h>
 
 /* Windows userland links with OpenSSL */
 #if !defined (_WIN32) || defined (_KERNEL)

--- a/module/icp/include/modes/gcm_asm_rename_funcs.h
+++ b/module/icp/include/modes/gcm_asm_rename_funcs.h
@@ -1,0 +1,39 @@
+// SPDX-License-Identifier: CDDL-1.0
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or https://opensource.org/licenses/CDDL-1.0.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2026 Attila Fülöp <attila@fueloep.org>
+ */
+
+/*
+ * Prepend `icp_` to each function name defined in gcm assembly files.
+ * This avoids potential symbol conflicts with linux libcrypto in case of
+ * in-tree compilation. To keep the diff noise low, we do this using macros.
+ *
+ * Currently only done for aesni-gcm-avx2-vaes.S since there is a real conflict.
+ */
+
+/* module/icp/asm-x86_64/modes/aesni-gcm-avx2-vaes.S */
+#define	gcm_init_vpclmulqdq_avx2	icp_gcm_init_vpclmulqdq_avx2
+#define	gcm_gmult_vpclmulqdq_avx2	icp_gcm_gmult_vpclmulqdq_avx2
+#define	gcm_ghash_vpclmulqdq_avx2	icp_gcm_ghash_vpclmulqdq_avx2
+#define	aes_gcm_enc_update_vaes_avx2	icp_aes_gcm_enc_update_vaes_avx2
+#define	aes_gcm_dec_update_vaes_avx2	icp_aes_gcm_dec_update_vaes_avx2

